test/jdk/java/util/IdentityHashMap/Basic.java

```diff
@@ -0,0 +1,546 @@
+        checkElements(keySet, k1a, k1b, k2);
+        assertTrue(keySet.equals(map2.keySet()));
+        assertTrue(map.equals(map2));
+    }
+
+    @Test
+    public void testKeySetRemove() {
+        Set<Box> keySet = map.keySet();
+        keySet.remove(k1a);
+        checkElements(keySet, k1b, k2);
+        checkEntries(map.entrySet(), entry(k1b, v1b),
+                                     entry(k2, v2));
+    }
+
+    @Test
+    public void testValues() {
+        Collection<Box> values = map.values();
+        checkElements(values, v1a, v1b, v2);
+        assertFalse(values.contains(new Box(v1a)));
+    }
+
+    @Test
+    public void testValuesNoRemove() {
+        Collection<Box> values = map.values();
+        values.remove(new Box(v1a));
+        checkElements(values, v1a, v1b, v2);
+        assertTrue(map.equals(map2));
+    }
+
+    @Test
+    public void testValuesRemove() {
+        Collection<Box> values = map.values();
+        values.remove(v1a);
+        checkElements(values, v1b, v2);
+        checkEntries(map.entrySet(), entry(k1b, v1b),
+                                     entry(k2, v2));
+    }
+
+    @Test
+    public void testEntrySet() {
+        Set<Map.Entry<Box,Box>> entrySet = map.entrySet();
+
+        assertFalse(entrySet.contains(entry(new Box(k1a), v1a)));
+        assertFalse(entrySet.contains(entry(k1b, new Box(v1b))));
+        assertFalse(entrySet.contains(entry(new Box(k2), new Box(v2))));
+
+        checkEntries(entrySet, entry(k1a, v1a),
+                               entry(k1b, v1b),
+                               entry(k2, v2));
+    }
+
+    @Test
+    public void testEntrySetNoRemove() {
+        Set<Map.Entry<Box, Box>> entrySet = map.entrySet();
+        entrySet.remove(entry(new Box(k1a), v1a));
+        entrySet.remove(entry(k1a, new Box(v1a)));
+        assertTrue(entrySet.equals(map2.entrySet()));
+        assertTrue(map.equals(map2));
+        checkEntries(entrySet, entry(k1a, v1a),
+                               entry(k1b, v1b),
+                               entry(k2, v2));
+    }
+
+    @Test
+    public void testEntrySetRemove() {
+        Set<Map.Entry<Box, Box>> entrySet = map.entrySet();
+        entrySet.remove(Map.entry(k1a, v1a));
+
+        checkEntries(entrySet, entry(k1b, v1b),
+                               entry(k2, v2));
+
+        checkEntries(map.entrySet(), entry(k1b, v1b),
+                                     entry(k2, v2));
+    }
+
+    // put
+    @Test
+    public void testPutNew() {
+        Box newKey = new Box(k1a);
+        Box newVal = new Box(v1a);
+        map.put(newKey, newVal);
+
+        checkEntries(map.entrySet(), entry(k1a, v1a),
+                                     entry(k1b, v1b),
+                                     entry(k2, v2),
+                                     entry(newKey, newVal));
+    }
+
+    // put
+    @Test
+    public void testPutOverwrite() {
+        Box newVal = new Box(v1a);
+        map.put(k1a, newVal);
```

Comment:
Should we capture the return value and assert that it's not null and is identity equal to `v1a`?

Reply:
Yes, good point, I'll add checks of the return values in the appropriate places. There are several, including `remove`, `computeX`, `put`, etc.